src/ast-parser.y

```diff
@@ -635,55 +642,65 @@ expr1 :
   | BLOCK labeling_opt block {
       Expr* expr = new_block_expr();
       expr->block = $3;
-      expr->block.label = $2;
+      expr->block->label = $2;
       $$ = join_exprs1(&@1, expr);
     }
   | LOOP labeling_opt block {
       Expr* expr = new_loop_expr();
       expr->loop = $3;
-      expr->loop.label = $2;
+      expr->loop->label = $2;
       $$ = join_exprs1(&@1, expr);
     }
   | IF labeling_opt value_type_list if_ {
       $$ = $4;
       Expr* if_ = $4.last;
       assert(if_->type == ExprType::If);
-      if_->if_.true_.label = $2;
-      if_->if_.true_.sig = $3;
+      if_->if_.true_->label = $2;
+      if_->if_.true_->sig = std::move(*$3);
+      delete $3;
     }
 ;
 if_ :
     LPAR THEN instr_list RPAR LPAR ELSE instr_list RPAR {
       Expr* expr = new_if_expr();
-      expr->if_.true_.first = $3.first;
+      expr->if_.true_ = new Block();
+      expr->if_.true_->first = $3.first;
       expr->if_.false_ = $7.first;
       $$ = join_exprs1(&@1, expr);
     }
   | LPAR THEN instr_list RPAR {
       Expr* expr = new_if_expr();
-      expr->if_.true_.first = $3.first;
+      expr->if_.true_ = new Block();
+      expr->if_.true_->first = $3.first;
+      expr->if_.false_ = nullptr;
       $$ = join_exprs1(&@1, expr);
     }
   | expr LPAR THEN instr_list RPAR LPAR ELSE instr_list RPAR {
       Expr* expr = new_if_expr();
-      expr->if_.true_.first = $4.first;
+      expr->if_.true_ = new Block();
+      expr->if_.true_->first = $4.first;
       expr->if_.false_ = $8.first;
       $$ = join_exprs2(&@1, &$1, expr);
     }
   | expr LPAR THEN instr_list RPAR {
       Expr* expr = new_if_expr();
-      expr->if_.true_.first = $4.first;
+      expr->if_.true_ = new Block();
+      expr->if_.true_->first = $4.first;
+      expr->if_.false_ = nullptr;
       $$ = join_exprs2(&@1, &$1, expr);
     }
   | expr expr expr {
       Expr* expr = new_if_expr();
-      expr->if_.true_.first = $2.first;
+      expr->if_.true_ = new Block();
+      expr->if_.true_->first = $2.first;
       expr->if_.false_ = $3.first;
       $$ = join_exprs2(&@1, &$1, expr);
     }
   | expr expr {
       Expr* expr = new_if_expr();
-      expr->if_.true_.first = $2.first;
+      expr->if_.true_ = new Block();
+      expr->if_.true_->first = $2.first;
+      expr->if_.false_ = nullptr;
```

Comment:
Should the initialization of field false_ start with the nullptr, removing the need for this explicit assignment?